### PR TITLE
Update to loopback 2

### PR DESCRIPTION
**Pull request by hacksparrow**
Updates the project to use `loopback` 2 and its conventions.

### CODE REVIEW

**Comment by hacksparrow** [comment]
@crandmck please don't merge this yet. @bajtos wanted to make some more updates to the examples. Will get them done and let you know.

**Comment by bajtos** [comment]
@hacksparrow this patch LGTM, what are the additional updates you mentioned in your comment?

**Comment by hacksparrow** [comment]
@bajtos the loopback#middleware pattern and `server/config.json` format.

**Comment by hacksparrow** [comment]
@bajtos updating `explorer` causes this to be printed:
```
Swagger: skipping unknown type "GeoPoint".
```
Any idea what's going and what could be the effect? The example seems to be working fine, though.

**Comment by hacksparrow** [comment]
@bajtos @crandmck, all updated. Should be good to merge now.

**Comment by bajtos** [comment]
> Swagger: skipping unknown type "GeoPoint".

See https://github.com/strongloop/loopback-explorer/issues/116

> the loopback#middleware pattern and server/config.json format.

Ah, good catch, I forgot about these.

File: package.json [hunk 1]
```diff
@@ -1,25 +1,27 @@
...
-    "loopback-datasource-juggler": "^2.0.0",
-    "serve-favicon": "^2.0.1",
     "function-rate-limit": "~0.0.1",
+    "loopback": "^2.22.0",
+    "loopback-boot": "^2.12.1",
+    "loopback-connector-rest": "^1.1.4",
+    "loopback-datasource-juggler": "^2.0.0",
     "request": "^2.37.0",
+    "serve-favicon": "^2.0.1",
+    "serve-static": "^1.10.0",
```
Comment (bajtos) [inline package.json]: FWIW, `loopback#static` remains a supported middleware, it's not necessary to use `server-static` directly. But then there is no harm in that either.

File: package.json [hunk 2]
```diff
@@ -1,25 +1,27 @@
...
-    "loopback-boot": "^2.0.0",
-    "loopback-datasource-juggler": "^2.0.0",
-    "serve-favicon": "^2.0.1",
     "function-rate-limit": "~0.0.1",
+    "loopback": "^2.22.0",
+    "loopback-boot": "^2.12.1",
+    "loopback-connector-rest": "^1.1.4",
+    "loopback-datasource-juggler": "^2.0.0",
     "request": "^2.37.0",
+    "serve-favicon": "^2.0.1",
```
Comment (bajtos) [inline package.json]: I believe this dependency is not necessary, loopback already includes `favicon` in its dependencies.

File: package.json [hunk 3]
```diff
@@ -1,25 +1,27 @@
...
     "shuffle": "~0.2.1",
     "text-table": "~0.2.0",
-    "weighted": "~0.2.2",
-    "async": "~0.9.0",
-    "loopback-connector-rest": "^1.1.4"
+    "weighted": "~0.2.2"
   },
   "optionalDependencies": {
-    "loopback-explorer": "^1.1.0",
+    "loopback-explorer": "^2.0.0",
```
Comment (bajtos) [inline package.json]: Explorer is not an optional dependency anymore, please move it to regular dependencies.

File: server/middleware.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  "files": {
+    "serve-static": {
+      "params": "$!../client"
+    }
+  },
+  "final": {
+    "loopback#urlNotFound": {}
+  },
+  "final:after": {
+    "errorhandler": {}
```
Comment (bajtos) [inline server/middleware.json]: Please use `loopback#errorHandler`, it adds extra features on top of vanilla `errorhandler` - see https://github.com/strongloop/loopback/pull/1502.

(the docs were not updated yet)

/cc @crandmck

File: package.json (hunk 1 above)
Comment (hacksparrow) [inline package.json]: I think it is good to start phasing out `loopback#middleware` which don't modify the underlying middleware.

**Comment by hacksparrow** [comment]
@bajtos made the changes you suggested. Jenkins is failing without any specifics, though.

**Comment by hacksparrow** [comment]
Jenkins is fine now.

@bajtos @crandmck ready to merge?

**Comment by crandmck** [comment]
LGTM.  If you've included all of Miroslav's suggestions, please merge.

**Comment by hacksparrow** [comment]
@crandmck I don't seem to have merge access. @bajtos can you?

**Comment by crandmck** [comment]
Since Miro. is probably asleep, I merged it. :-)

**Comment by hacksparrow** [comment]
👍

File: package.json (hunk 1 above)
Comment (bajtos) [inline package.json]: Makes sense, although `loopback#static` is coming straight from express. IIRC serve-static is the only middleware bundled in express@4.

**Pull request merged**